File: AdventOfCode2023/day1-2.go
package main

import (
	"bufio"
	"os"
	"strings"
	"unicode"
)

func strToNum(s string) int {
	if strings.HasPrefix(s, "one") {
		return 1
	} else if strings.HasPrefix(s, "two") {
		return 2
	} else if strings.HasPrefix(s, "three") {
		return 3
	} else if strings.HasPrefix(s, "four") {
		return 4
	} else if strings.HasPrefix(s, "five") {
		return 5
	} else if strings.HasPrefix(s, "six") {
		return 6
	} else if strings.HasPrefix(s, "seven") {
		return 7
	} else if strings.HasPrefix(s, "eight") {
		return 8
	} else if strings.HasPrefix(s, "nine") {
		return 9
	}

	return 0
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	vals := make([]int, 0)
	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			break
		}

		first := -1
		last := -1
		for i, ch := range line {
			if unicode.IsDigit(rune(ch)) || strToNum(line[i:]) != 0 {
				n := isNum(line[i:])
				if n == 0 {
					n = int(ch - rune('0'))
				}
				if first == -1 {
					first = n
				}
				last = n
			}
		}
		first *= 10
		first += last
		vals = append(vals, first)
	}

	sum := 0
	for _, v := range vals {
		sum += v
	}
	println("Ans:", sum)
}

File: AdventOfCode2023/day1-2.rs
use std::io;

fn main() {
    let mut ans = 0;
    for line in io::stdin().lines() {
        let mut string = line.unwrap();
        string = string.replace("one", "o1e");
        string = string.replace("two", "t2o");
        string = string.replace("three", "th3ee");
        string = string.replace("four", "f4ur");
        string = string.replace("five", "fi5e");
        string = string.replace("six", "s6x");
        string = string.replace("seven", "se7en");
        string = string.replace("eight", "ei8ht");
        string = string.replace("nine", "ni9e");

        let mut first: i32 = -1;
        let mut last: i32 = -1;
        for ch in string.chars() {
            if ch.is_digit(10) {
                if first == -1 {
                    first = ch.to_digit(10).unwrap() as i32;
                }
                last = ch.to_digit(10).unwrap() as i32;
            }
        }
        ans += (first*10)+last;
    }
    println!("Ans: {ans}")
}
